packages/@react-stately/datepicker/src/useDateRangePickerState.ts

```diff
@@ -42,16 +47,53 @@ export function useDateRangePickerState(props: DateRangePickerProps): DateRangeP
     onChange
   );
 
+  let v = (value?.start || value?.end || props.placeholderValue);
+  let granularity = props.granularity || (v && 'minute' in v ? 'minute' : 'day');
+  let hasTime = granularity === 'hour' || granularity === 'minute' || granularity === 'second' || granularity === 'millisecond';
+
+  let [dateRange, setSelectedDateRange] = useState<DateRange>(null);
+  let [timeRange, setSelectedTimeRange] = useState<TimeRange>(null);
+
+  if (value && value.start && value.end) {
+    dateRange = value;
+    if ('hour' in value.start) {
+      // @ts-ignore
+      timeRange = value;
```

Comment:
```suggestion
      timeRange = value as DateRange;
```
I think it's the guarding against null that we're ignoring? but at this point with the if check we know it's not null anymore